### PR TITLE
fix($compile): do not swallow thrown errors in test

### LINKED ISSUE #15629: missing required controllers doesn't result in an exception in tests
**Do you want to request a *feature* or report a *bug*?**
bug

**What is the current behavior?**
neither the $compile() or subsequent $digest() calls will throw an exception when using a directive with an unsatisfied "require" requirement.  NOTE: This is a change from 1.5.10 to 1.6.x

Plunkr: http://plnkr.co/edit/YaJbS1?p=preview

**What is the expected behavior?**
I expect it to throw when something that's required is not present

**What is the motivation / use case for changing the behavior?**
It used to work that way and the directive documentation says that $compile with throw (though it takes an extra digest to throw when you use a templateUrl - also doesn't throw if you don't have a 'link' method)

**Which versions of Angular, and which browser / OS are affected by this issue? Did this work in previous versions of Angular? Please also test with the latest stable and snapshot (https://code.angularjs.org/snapshot/) versions.**
All Versions 1.6.x

**Other information (e.g. stacktraces, related issues, suggestions how to fix)**
In my plunker, you can copy the context of "index-161.html" to "index.html" to test 1.6.1, and "index-1510.html" to test 1.5.10

### PR DESCRIPTION
**Pull request by gkalpak**
**What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)**
Bug fix.

**What is the current behavior? (You can also link to an open issue here)**
Since e13eeab, errors/rejections produces during fetching the template or compiling an asynchronous directive, where overzealously silenced. This doesn't make any difference in (most) production apps, where `$exceptionHandler` does not rethrow the errors. In tests though (where `$exceptionHandler` rethrows by default), it can unexpectedly "swallow" thrown errors.
See also #15629.

**What is the new behavior (if this is a feature change)?**
The extraneous `.catch(noop)` has been removed, thus letting errors thrown by `$exceptionHandler` to surface.

**Does this PR introduce a breaking change?**
No
<sub>(It is possible that exceptions might pop up in tests, but they where previously incorrectly hidden. Also note that the only modification in tests were to restore them to their pre-1.6.0-rc.0 state.)</sub>

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md#commit-message-format
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] ~~Docs have been added / updated (for bug fixes / features)~~

**Other information**:
The changes in 'compileSpec.js' essentially revert the modifications that were unnecessarily (and incorrectly) done in e13eeab (and also one incorrect modification from [c22615c][1]).

Fixes #15629

[1]: https://github.com/angular/angular.js/commit/c22615cbfbaa7d1712e79b6bf2ace6eb41313bac#diff-348c2f3781ed66a24894c2046a52c628L2084

### CODE REVIEW

**Comment by lgalfaso** [comment]
Is this a breaking change? Even when this is only for tests (and I am ok in
having it in the current stable branch).

**Comment by gkalpak** [comment]
> Is this a breaking change?

It depends how you look at it :grin:

First of all, this will have no effect in cases where `$exceptionHandler` does not throw an exception. In most production apps this is not the case (and if it is, people will run into bigger problems, such as half-finished digest cycles and inconsistent state).

In cases where `$exceptionHandler` rethrows (the default behavior in tests), this PR will restore the 1.5.x behavior (throwing), which was unintentionally changed/broken in e13eeab (and started swallowing errors).

**tl;dr**
Theoretically, this PR fixes a regression introduced by e13eeab in 1.6.0-rc.0.
Practically, it is possible that people have written tests that pass with 1.6.0-rc.0+ but fail with this PR.

**Comment by lgalfaso** [comment]
I am fine with the explanation for not calling this a BR. LGTM

**Comment by Narretz** [comment]
LGTM
